interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/SpelExpressionSuggester.scala

```diff
@@ -0,0 +1,154 @@
+      insertDummyVariable(spelExpression, normalizedCaretPosition)
+    } else {
+      spelExpression
+    }
+
+    def filterMapByName[V](map: Map[String, V], name: String): Map[String, V] = {
+      if (shouldInsertDummyVariable) {
+        map
+      } else {
+        map.filter { case (key, _) => key.toLowerCase.contains(name.toLowerCase) }
+      }
+    }
+
+    def suggestionsForPropertyOrFieldReference(nodeInPosition: NuSpelNode, p: PropertyOrFieldReference): Future[Iterable[ExpressionSuggestion]] = {
+      val typedPrevNode = nodeInPosition.prevNode().flatMap(_.typingResultWithContext)
+      typedPrevNode.collect {
+        case TypingResultWithContext(tc: TypedClass, staticContext) => Future.successful(typeDefinitions.get(tc.klass).map(c => filterClassMethods(c, p.getName, staticContext)).getOrElse(Nil))
+        case TypingResultWithContext(to: TypedObjectWithValue, staticContext) => Future.successful(typeDefinitions.get(to.underlying.klass).map(c => filterClassMethods(c, p.getName, staticContext)).getOrElse(Nil))
+        case TypingResultWithContext(to: TypedObjectTypingResult, _) =>
+          val suggestionsFromFields = filterMapByName(to.fields, p.getName).toList.map { case (methodName, clazzRef) => ExpressionSuggestion(methodName, clazzRef, fromClass = false) }
+          val suggestionsFromClass = typeDefinitions.get(to.objType.klass).map(c => filterClassMethods(c, p.getName, staticContext = false, fromClass = suggestionsFromFields.nonEmpty)).getOrElse(Nil)
+          Future.successful(suggestionsFromFields ++ suggestionsFromClass)
+        case TypingResultWithContext(tu: TypedUnion, staticContext) => Future.successful(tu.possibleTypes.map(_.objType.klass).flatMap(klass => typeDefinitions.get(klass).map(c => filterClassMethods(c, p.getName, staticContext)).getOrElse(Nil)))
+        case TypingResultWithContext(td: TypedDict, _) => dictQueryService.queryEntriesByLabel(td.dictId, if (shouldInsertDummyVariable) "" else p.getName)
+          .map(_.map(list => list.map(e => ExpressionSuggestion(e.label, td, fromClass = false)))).getOrElse(successfulNil)
+      }.getOrElse(successfulNil)
+    }
+
+    def filterClassMethods(classDefinition: ClazzDefinition, name: String, staticContext: Boolean, fromClass: Boolean = false): List[ExpressionSuggestion] = {
+      val methods = filterMapByName(if(staticContext) classDefinition.staticMethods else classDefinition.methods, name)
+
+      methods.values.flatten
+        .map(m => ExpressionSuggestion(m.name, m.signatures.head.result, fromClass = fromClass))
+        .toList
+    }
+
+    val suggestions = for {
+      parsedSpelNode <- nuSpelNodeParser.parse(input, variables).toOption
+      nodeInPosition <- parsedSpelNode.findNodeInPosition(normalizedCaretPosition)
+    } yield {
+      nodeInPosition.spelNode match {
+        // variable is typed (#foo), so we need to return filtered list of all variables that match currently typed name
+        case v: VariableReference =>
```

Comment:
I see one import, missing feature here. You don't suggest `#this` inside projection and selection. But we can move this to the next PR to not make this PR growth too much.

Reply:
Will be done in next PR